File: scissari-executor-fix/scissari_executor/service.py
"""ExecutorRunService (Facade).

The ONLY entry point Scissari's turn loop calls. Orchestrates:

    breaker.allow() -> client.run() -> classifier.classify()
      -> factory.for_kind().recover() -> guard.register() -> alert_sink.emit()

All collaborators are injected (Dependency Injection) so the facade can be
unit-tested with fakes. On any unrecoverable path it emits a StallReport that
finally carries a concrete reason and raises StalledError.
"""
from __future__ import annotations

import asyncio
from typing import Optional

from .interfaces import (
    IExecutorRunService,
    IExecutorClient,
    IFailureClassifierChain,
    IRecoveryStrategyFactory,
    ILoopGuard,
    ICircuitBreaker,
    IAlertSink,
    IConversationSnapshotStore,
)
from .classifiers import build_default_chain
from .strategies import StrategyFactory
from .guard import LoopGuard
from .breaker import CircuitBreaker
from .models import (
    ExecutorCommand,
    ExecutorResponse,
    ExecutorFailureError,
    FailureClassification,
    FailureKind,
    RecoveryAction,
    RecoveryOutcome,
    StallReport,
    TurnState,
    GuardVerdict,
)


class StalledError(RuntimeError):
    """Raised when the turn cannot complete. Carries the StallReport."""

    def __init__(self, report: StallReport):
        self.report = report
        super().__init__(report.message)


# Hard ceiling so a logic bug can never reproduce the original infinite spin.
_ABSOLUTE_MAX_ATTEMPTS = 6


class ExecutorRunService(IExecutorRunService):
    def __init__(
        self,
        client: IExecutorClient,
        alert_sink: IAlertSink,
        classifier: Optional[IFailureClassifierChain] = None,
        factory: Optional[IRecoveryStrategyFactory] = None,
        guard: Optional[ILoopGuard] = None,
        breaker: Optional[ICircuitBreaker] = None,
        snapshot_store: Optional[IConversationSnapshotStore] = None,
        sleep=asyncio.sleep,
    ):
        self._client = client
        self._alert_sink = alert_sink
        self._classifier = classifier or build_default_chain()
        self._factory = factory or StrategyFactory()
        self._guard = guard or LoopGuard(snapshot_store=snapshot_store)
        self._breaker = breaker or CircuitBreaker()
        self._snapshots = snapshot_store
        self._sleep = sleep

    async def _stall(
        self,
        agent_id: str,
        classification: Optional[FailureClassification],
        verdict: Optional[GuardVerdict],
        note: str,
    ) -> "StalledError":
        report = StallReport(
            agent_id=agent_id,
            classification=classification,
            calls_used=verdict.calls_used if verdict else 0,
            final_state=verdict.state if verdict else TurnState.TRIPPED,
            snapshot_id=verdict.snapshot_id if verdict else None,
            # The fix: a concrete, classified reason — never "no error captured".
            message=(
                f"executor_run could not complete: {note}"
                + (f" — {classification.evidence}" if classification else "")
            ),
        )
        await self._alert_sink.emit(report)
        return StalledError(report)

    async def execute(self, cmd: ExecutorCommand, agent_id: str) -> ExecutorResponse:
        current = cmd

        for _ in range(_ABSOLUTE_MAX_ATTEMPTS):
            # Circuit breaker: don't even touch a service we believe is dead.
            if not self._breaker.allow():
                classification = FailureClassification(
                    kind=FailureKind.EXECUTOR_DOWN,
                    retryable=False,
                    recommended_action=RecoveryAction.CIRCUIT_OPEN,
                    evidence="circuit open — executor presumed down",
                    classifier_name="circuit_breaker",
                )
                verdict = self._guard.register(
                    current,
                    RecoveryOutcome(
                        action=RecoveryAction.CIRCUIT_OPEN,
                        kind=classification.kind,
                        reason="circuit open",
                    ),
                )
                raise await self._stall(
                    agent_id, classification, verdict, "circuit open"
                )

            try:
                response = await self._client.run(current)
            except ExecutorFailureError as exc:
                classification = self._classifier.classify(exc.failure)
                self._breaker.record_failure(classification.kind)

                strategy = self._factory.for_kind(classification.kind)
                outcome = await strategy.recover(current, classification)
                outcome.kind = classification.kind  # tag for guard budget + reporting

                verdict = self._guard.register(current, outcome)

                # Terminal action or guard tripped -> stop and alert with a reason.
                if (
                    outcome.action in (RecoveryAction.ABORT, RecoveryAction.CIRCUIT_OPEN)
                    or not verdict.should_continue
                ):
                    raise await self._stall(
                        agent_id, classification, verdict, outcome.action.value
                    )

                # Retry-ish: honor backoff, advance the command, loop again.
                if outcome.backoff_ms:
                    await self._sleep(outcome.backoff_ms / 1000)
                current = outcome.next_command or current
                continue
            else:
                self._breaker.record_success()
                return response

        # Defensive: absolute attempt ceiling reached (should be unreachable).
        raise await self._stall(
            agent_id, None, None, "attempt ceiling reached"
        )

File: src/executor-stall-recovery/service.ts
/**
 * ExecutorRunService (Facade).
 * The ONLY entry point the bot's turn loop calls. Orchestrates all recovery logic.
 */

import { CircuitBreaker } from "./breaker";
import { buildDefaultChain } from "./classifiers";
import { LoopGuard } from "./guard";
import type {
  IAlertSink,
  ICircuitBreaker,
  IConversationSnapshotStore,
  IExecutorClient,
  IExecutorRunService,
  IFailureClassifierChain,
  ILoopGuard,
  IRecoveryStrategyFactory,
} from "./interfaces";
import type {
  ExecutorCommand,
  ExecutorResponse,
  FailureClassification,
  GuardVerdict,
  RecoveryOutcome,
} from "./models";
import {
  ExecutorFailureError,
  FailureKind,
  RecoveryAction,
  type StallReport,
  TurnState,
} from "./models";
import { StrategyFactory } from "./strategies";

// Hard ceiling so a logic bug can never reproduce the original infinite spin.
const ABSOLUTE_MAX_ATTEMPTS = 6;

export class StalledError extends Error {
  report: StallReport;

  constructor(report: StallReport) {
    super(report.message);
    this.name = "StalledError";
    this.report = report;
  }
}

export class ExecutorRunService implements IExecutorRunService {
  private client: IExecutorClient;
  private alertSink: IAlertSink;
  private classifier: IFailureClassifierChain;
  private factory: IRecoveryStrategyFactory;
  private guard: ILoopGuard;
  private breaker: ICircuitBreaker;
  private snapshots: IConversationSnapshotStore | null;
  private sleepFn: (ms: number) => Promise<void>;

  constructor(
    client: IExecutorClient,
    alertSink: IAlertSink,
    classifier?: IFailureClassifierChain,
    factory?: IRecoveryStrategyFactory,
    guard?: ILoopGuard,
    breaker?: ICircuitBreaker,
    snapshotStore?: IConversationSnapshotStore,
    sleepFn?: (ms: number) => Promise<void>,
  ) {
    this.client = client;
    this.alertSink = alertSink;
    this.classifier = classifier || buildDefaultChain();
    this.factory = factory || new StrategyFactory();
    this.guard = guard || new LoopGuard(undefined, snapshotStore);
    this.breaker = breaker || new CircuitBreaker();
    this.snapshots = snapshotStore || null;
    this.sleepFn =
      sleepFn || ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
  }

  private async stall(
    agentId: string,
    classification: FailureClassification | null,
    verdict: GuardVerdict | null,
    note: string,
  ): Promise<StalledError> {
    const report: StallReport = {
      agent_id: agentId,
      classification: classification || undefined,
      calls_used: verdict?.calls_used || 0,
      final_state: verdict?.state || TurnState.TRIPPED,
      snapshot_id: verdict?.snapshot_id,
      // The fix: a concrete, classified reason — never "no error captured".
      message:
        `executor_run could not complete: ${note}` +
        (classification ? ` — ${classification.evidence}` : ""),
    };
    await this.alertSink.emit(report);
    return new StalledError(report);
  }

  async execute(
    cmd: ExecutorCommand,
    agentId: string,
  ): Promise<ExecutorResponse> {
    let current = cmd;

    for (let attempt = 0; attempt < ABSOLUTE_MAX_ATTEMPTS; attempt++) {
      // Circuit breaker: don't even touch a service we believe is dead.
      if (!this.breaker.allow()) {
        const classification: FailureClassification = {
          kind: FailureKind.EXECUTOR_DOWN,
          retryable: false,
          recommended_action: RecoveryAction.CIRCUIT_OPEN,
          evidence: "circuit open — executor presumed down",
          classifier_name: "circuit_breaker",
        };
        const verdict = this.guard.register(current, {
          action: RecoveryAction.CIRCUIT_OPEN,
          kind: classification.kind,
          reason: "circuit open",
        });
        throw await this.stall(
          agentId,
          classification,
          verdict,
          "circuit open",
        );
      }

      try {
        const response = await this.client.run(current);
        this.breaker.record_success();
        return response;
      } catch (error) {
        if (!(error instanceof ExecutorFailureError)) {
          throw error;
        }

        const classification = this.classifier.classify(error.failure);
        this.breaker.record_failure(classification.kind);

        const strategy = this.factory.for_kind(classification.kind);
        const outcome = await strategy.recover(current, classification);
        outcome.kind = classification.kind; // tag for guard budget + reporting

        const verdict = this.guard.register(current, outcome);

        // Terminal action or guard tripped -> stop and alert with a reason.
        if (
          [RecoveryAction.ABORT, RecoveryAction.CIRCUIT_OPEN].includes(
            outcome.action,
          ) ||
          !verdict.should_continue
        ) {
          throw await this.stall(
            agentId,
            classification,
            verdict,
            outcome.action,
          );
        }

        // Retry-ish: honor backoff, advance the command, loop again.
        if (outcome.backoff_ms) {
          await this.sleepFn(outcome.backoff_ms);
        }
        current = outcome.next_command || current;
      }
    }

    // Defensive: absolute attempt ceiling reached (should be unreachable).
    throw await this.stall(agentId, null, null, "attempt ceiling reached");
  }
}
